### ISSUE TITLE
Benchmark feasibility/performance of `score_genes` algorithm for coloring by gene sets

### ISSUE THREAD

**Issue by sidneymbell** [chanzuckerberg/cellxgene#949]
Function in scanpy: https://github.com/theislab/scanpy/blob/master/scanpy/tools/_score_genes.py#L14

Original paper: 
https://www.nature.com/articles/nbt.3192.pdf

**Comment by bkmartinjr** [comment]
Related issue: this function fails when run on file-backed data.  See theislab/scanpy#860   Unlikely we can support scanpy functions until they correctly work with all supported datatypes and reading modes (dense, CSC, CSR, backed/memory).

**Comment by bkmartinjr** [comment]
Initial evaluation is based on visually inspection and benchmarking of the scanpy implementation.   It does not appear to be an algorithm suitable for interactive use on anything other than small (< 10K cell) datasets, assuming we use the default parameters.  

By default, the algorithm scores a set of genes (`gene_list`) against a reference set of other genes (`gene_pool`), the latter which defaults to all genes.   In the scanpy implementation, this will densify the entire X sub-matrix for those reference genes, ie, by default the entire X matrix.   Even for moderately large datasets, this will be larger than the RAM available on a typical laptop.  For example, the Broad immune cell test dataset is approximately [380000 x 20000] or ~30GB, and we have users with larger datasets.

The critical section of the Scanpy implementation:
```
# TODO: this densifies the whole data matrix for `gene_pool`
if issparse(_adata.X):
    obs_avg = pd.Series(
        np.nanmean(
            _adata[:, gene_pool].X.toarray(), axis=0), index=gene_pool)  # average expression of genes
else:
    obs_avg = pd.Series(
        np.nanmean(_adata[:, gene_pool].X, axis=0), index=gene_pool)  # average expression of genes
```

This mean expression value is used to create bins of cells with similar expression, from which a number of genes are randomly selected and used in scoring.  It _might_ be possible to approximate this process with less data by randomly sampling, rather than calculating the mean expression bins across all cells.  It would require some effort to determine the utility of this approach.

I believe (but have not tested) that it is possible to improve the above mean calculation by methods similar to our custom diffexp (eg. don't use np.nanmean, but instead use the sparse mean function).  This should be somewhat faster for CSC encoded data, but slower for CSR.

For datasets that do fit in memory, speed of the scanpy implementation is below, using default parameters (these are approximations, averaging among different sparse/dense encodings, different number of genes being scored, etc - none of these parameters grossly changed the result), and running on my laptop (fast Intel/32GB):
* 3K x 2K, pbmc3K dataset:  < 0.2 seconds
* 33K x 14K, pbmc33K dataset: 13-15 seconds
* 54K x 1K, TM_facs dataset: ~2 seconds

For sparse datasets larger than the above examples, my 32GB laptop ran out of memory due to the densification step, even though cellxgene can currently load/view them (eg, Regev Lab immune dataset).

In summary:
* with default parameters, which require computing stats across the entire matrix, interactive performance is not achievable with anything other than very small datasets.
* we could potentially improve performance somewhat by modifying the mean calculation used to bin reference genes by like expression
* all performance is highly subject to the size of the reference gene set (gene_pool) and other params.  

Suggest we look for algorithms that use var less data (eg, sampling) or a means of pre-selecting a smaller reference gene set.

**Comment by fidelram** [comment]
The score genes function from scanpy is a bit old compared to newer developments. If I remember correctly, @LuckyMD uses a different method to score genes (and assign them to cell types) in his best practices notebook (https://github.com/theislab/single-cell-tutorial/blob/master/latest_notebook/Case-study_Mouse-intestinal-epithelium_1906.ipynb, see: `sc.tl.marker_gene_overlap`)

A problem with the methods to score genes is that they usually require normalized expression values that tend to densify the matrix.  This means that other methods may have the same problem.

**Comment by bkmartinjr** [comment]
Thanks @fidelram. If you know of useful scoring methods that are potentially interactive speed (a few seconds for 1M x 30K data set), could you let us know?  My understanding is that the general feature is desired, not necessarily a particular algorithm.

**Comment by LuckyMD** [comment]
I would assume that once you have data-driven marker genes (`sc.tl.rank_genes_groups`), marker gene overlaps should be the fastest way to go. I'm sure you can speed up my marker gene overlap function for that end.

**Comment by sidneymbell** [comment]
Thanks so much for benchmarking, @bkmartinjr! This makes complete sense. 

I think it would be possible to use some clever random sampling to ease the computation of the `gene_pool` mean. But yes, you are correct that the desire is simply for a generalizable way to color by expression of an arbitrary set of genes. 

@LuckyMD & @fidelram -- Thank you for looking at this! I agree that this method is a bit outdated for some applications; I also prefer `marker_gene_overlap` for inspecting annotations. The tricky thing here is that we're looking to color by any arbitrary set of genes, not necessarily cell type markers. Any suggestions would be super helpful!

**Comment by LuckyMD** [comment]
`sc.tl.score_genes()` takes the average z-score of a set of genes and subtracts the average z-score of a random set. You could speed up the function by skipping the second part (random genes z-score subtraction) of that?
